gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.retention;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.util.Progressable;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import gobblin.runtime.Limiter;
+import gobblin.runtime.RateBasedLimiter;
+import gobblin.util.Decorator;
+
+
+/**
+ * Subclass of {@link org.apache.hadoop.fs.FileSystem} that wraps with a {@link gobblin.runtime.Limiter}
+ * to control HDFS call rate.
+ *
+ *  <p>
+ *  This classes uses Guava's {@link Cache} for storing {@link org.apache.hadoop.fs.FileSystem} URI to
+ *  {@link gobblin.runtime.Limiter} mapping.
+ *  </p>
+ *
+ *  <p>
+ *  For methods that require HDFS calls, this class will first acquire a permit using {@link gobblin.runtime.Limiter},
+ *  to make sure HDFS call rate is allowed by the uppper limit.
+ *  </p>
+ */
+public class RateControlledFileSystem extends FileSystem implements Closeable, Decorator {
+
+  private static final int DEFAULT_MAX_CACHE_SIZE = 100;
+  private static final Cache<String, Limiter> FS_URI_TO_RATE_LIMITER_CACHE = CacheBuilder.newBuilder()
```

Comment:
It's better to use a `CacheLoader` and initialize this in the constructor instead of creating a `Callable` every time `get` is called in `getRateLimiter`.